deep_quoridor/src/agents/simple.py

```diff
@@ -1,61 +1,175 @@
-from quoridor_env import StepRewardCalculator
+import numpy as np
+from quoridor import Action, ActionEncoder, Player, Quoridor, WallAction, construct_game_from_observation
+from utils import SubargsBase
 
 from agents.core import Agent
 
+# We use a large reward to encourage the agent to win the game, but we can't use infinity
+# because then multiplying by a discount factor won't decrease the reward.
+WINNING_REWARD = 1e6
 
-def sample_random_action_sequence(game: Quoridor, player: Player, opponent: Player, max_path_length: int):
+
+@dataclass
+class SimpleParams(SubargsBase):
+    nick: Optional[str] = None
+
+    # How many moves to look ahead in the minimax algorithm.
+    max_depth: int = 2
+
+    # How many actions to consider at each minimax stage.
+    branching_factor: int = 20
+
+    # We use a Guassian distribution centered around each player's position to sample wall actions.
+    # Sigma is the standard deviation of this distribution; a smaller sigma means the agent is more
+    # likely to place walls near themselves or their opponent, rather than somewhere else on the board.
+    wall_sigma: float = 0.5
+
+    # Future rewards are worth less. This makes the agent try to win quickly. Without this,
+    # once the agent knows it can win, it may oscillate between two moves instead of just winning.
+    discount_factor: float = 0.99
+
+
+def compute_wall_weight(wall: WallAction, position_1: np.ndarray, position_2: np.ndarray, sigma) -> float:
+    wall_position = np.array(wall.position)
+    d1 = np.linalg.norm(position_1 - wall_position)
+    d2 = np.linalg.norm(position_2 - wall_position)
+    min_distance = min(d1, d2)
+    weight = np.exp(-0.5 * (min_distance / sigma) ** 2)
+    return weight
+
+
+def sample_actions(game: Quoridor, n: int, wall_sigma: float | None = None) -> list[Action]:
+    """
+    Choose a sample of valid actions for the current player.
+
+    Starts by including all valid move actions, then adds a random sample of valid wall actions. May return less than
+    n actions if there are not enough valid actions.
     """
-    Sample a random sequence of actions for a given game. Stops early if the game terminates."""
-    action_sequence = []
-    total_reward = 0.0
-    reward_calculator = StepRewardCalculator(game, player, opponent)
-    while len(action_sequence) < max_path_length:
-        # For now, assume the other agent takes random actions.
-        valid_actions = game.get_valid_actions()
-        assert len(valid_actions) > 0, "No valid actions available... this shouldn't be possible."
+    actions = game.get_valid_move_actions()
+    if len(actions) > n:
+        actions = random.sample(actions, n)
+
+    if len(actions) < n:
+        num_wall_actions = n - len(actions)
+        wall_actions = game.get_valid_wall_actions()
+        if len(wall_actions) > num_wall_actions:
+            if wall_sigma is None:
+                wall_actions = random.sample(wall_actions, num_wall_actions)
```

Comment:
I think just picking up a random wall would never be a decent strategy, any reason to keep this?

Reply:
It's still sliiiightly useful so I'd like to keep it. I like having it as a comparison against the "choose walls near players" strategy. It definitely only works at all when the branching factor (and therefor num_wall_actions) is large. Increasing wall sigma to a very large value has a similar effect, but doesn't quite create a true uniform distribution.